File: _includes/sort-benchmark/algorithms/loser_tree.rs
fn player_key(a: &[usize], idx: usize) -> usize {
    if idx == usize::MAX {
        usize::MAX
    } else {
        a[idx]
    }
}

fn loser_and_winner(a: &[usize], left: usize, right: usize) -> (usize, usize) {
    match (left, right) {
        (usize::MAX, r) => (usize::MAX, r),
        (l, usize::MAX) => (usize::MAX, l),
        (l, r) => {
            if player_key(a, l) <= player_key(a, r) {
                (r, l)
            } else {
                (l, r)
            }
        }
    }
}

fn adjust_loser_tree(a: &[usize], ls: &mut [usize], k: usize, mut s: usize) {
    let mut t = (s + k) / 2;
    while t > 0 {
        if player_key(a, s) > player_key(a, ls[t]) {
            core::mem::swap(&mut s, &mut ls[t]);
        }
        t /= 2;
    }
    ls[0] = s;
}

fn loser_tree_sort(a: &mut [usize]) {
    let n = a.len();
    if n <= 1 {
        return;
    }
    let k = n.next_power_of_two();
    let mut ls = vec![usize::MAX; k];
    let mut winner = vec![usize::MAX; 2 * k];
    for i in 0..k {
        winner[k + i] = if i < n { i } else { usize::MAX };
    }
    for i in (1..k).rev() {
        let (loser, win) = loser_and_winner(a, winner[2 * i], winner[2 * i + 1]);
        ls[i] = loser;
        winner[i] = win;
    }
    ls[0] = winner[1];

    let mut out = vec![0usize; n];
    for pos in 0..n {
        let idx = ls[0];
        out[pos] = a[idx];
        a[idx] = usize::MAX;
        adjust_loser_tree(a, &mut ls, k, idx);
    }
    a.copy_from_slice(&out);
}

File: _includes/sort-benchmark/algorithms/loser_tree.swift
fileprivate func player_key(_ a: UnsafeMutableBufferPointer<Int>, _ idx: Int) -> Int {
    if idx == Int.max {
        return Int.max
    } else {
        return a[idx]
    }
}

fileprivate func next_power_of_two(_ n: UInt) -> UInt {
    if n <= 1 {
        return 1
    }
    var v = n - 1
    v |= v >> 1
    v |= v >> 2
    v |= v >> 4
    v |= v >> 8
    v |= v >> 16
    v |= v >> 32
    return v + 1
}

fileprivate func loser_and_winner(
    _ a: UnsafeMutableBufferPointer<Int>,
    _ left: Int,
    _ right: Int
) -> (Int, Int) {
    if left == Int.max {
        return (Int.max, right)
    }
    if right == Int.max {
        return (Int.max, left)
    }
    if player_key(a, left) <= player_key(a, right) {
        return (right, left)
    } else {
        return (left, right)
    }
}

fileprivate func adjust_loser_tree(
    _ a: UnsafeMutableBufferPointer<Int>,
    _ ls: inout [Int],
    _ k: Int,
    _ s: Int
) {
    var s = s
    var t = (s + k) / 2
    while t > 0 {
        if player_key(a, s) > player_key(a, ls[t]) {
            let tmp = s
            s = ls[t]
            ls[t] = tmp
        }
        t /= 2
    }
    ls[0] = s
}

func loser_tree_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { loser_tree_sort($0) }
}

func loser_tree_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    let n = a.count
    if n <= 1 {
        return
    }
    let k = Int(next_power_of_two(UInt(n)))
    var ls = [Int](repeating: Int.max, count: k)
    var winner = [Int](repeating: Int.max, count: 2 * k)
    for i in 0..<k {
        winner[k + i] = i < n ? i : Int.max
    }
    for i in (1..<k).reversed() {
        let (loser, win) = loser_and_winner(a, winner[2 * i], winner[2 * i + 1])
        ls[i] = loser
        winner[i] = win
    }
    ls[0] = winner[1]

    var out = [Int](repeating: 0, count: n)
    for pos in 0..<n {
        let idx = ls[0]
        out[pos] = a[idx]
        a[idx] = Int.max
        adjust_loser_tree(a, &ls, k, idx)
    }
    for i in 0..<n {
        a[i] = out[i]
    }
}
